### PR TITLE
add dev container schema

### PR DESCRIPTION
**Pull request by aeschli**
Implementing devcontainer schema proposal https://github.com/devcontainers/spec/issues/14#issuecomment-1185373811

### CODE REVIEW

File: schemas/devContainer.base.schema.json [hunk 1]
```diff
@@ -0,0 +1,576 @@
...
+						"type": "string"
+					}
+				},
+				"shutdownAction": {
+					"type": "string",
+					"enum": [
+						"none",
+						"stopContainer"
+					],
+					"description": "Action to take when the VS Code window is closed. The default is to stop the container."
```
Comment (jkeech) [inline schemas/devContainer.base.schema.json:368]:
```suggestion
					"description": "Action to take when the user disconnects from the container in their editor. The default is to stop the container."
```

File: schemas/devContainer.base.schema.json [hunk 2]
```diff
@@ -0,0 +1,576 @@
...
+						"array"
+					],
+					"description": "The name of the docker-compose file(s) used to start the services.",
+					"items": {
+						"type": "string"
+					}
+				},
+				"service": {
+					"type": "string",
+					"description": "The service you want to work on."
```
Comment (jkeech) [inline schemas/devContainer.base.schema.json:505]:
```suggestion
					"description": "The service you want to work on. This is considered the primary container for your dev environment which your editor will connect to."
```

File: schemas/devContainer.base.schema.json [hunk 3]
```diff
@@ -0,0 +1,576 @@
...
+					"type": "string",
+					"description": "The path of the workspace folder inside the container. This is typically the target path of a volume mount in the docker-compose.yml."
+				},
+				"shutdownAction": {
+					"type": "string",
+					"enum": [
+						"none",
+						"stopCompose"
+					],
+					"description": "Action to take when the VS Code window is closed. The default is to stop the containers."
```
Comment (jkeech) [inline schemas/devContainer.base.schema.json:524]:
```suggestion
					"description": "Action to take when the user disconnects from the primary container in their editor. The default is to stop all of the compose containers."
```

**Review by jkeech — COMMENTED** [review]
Thanks for adding these schema files! I left a few suggestions, including moving some VS Code and Codespaces-specific properties out of the root so that they are only documented under the `customizations` object, as well as a few other minor wording changes to the descriptions.

File: schemas/devContainer.base.schema.json [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+					"items": {
+						"type": "string",
+						"pattern": "^([a-z0-9A-Z][a-z0-9A-Z-]*)\\.([a-z0-9A-Z][a-z0-9A-Z-]*)((@(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?)|@prerelease)?$",
+						"errorMessage": "Expected format: '${publisher}.${name}' or '${publisher}.${name}@${version}'. Example: 'ms-dotnettools.csharp'."
+					}
+				},
+				"settings": {
+					"$ref": "vscode://schemas/settings/machine",
+					"description": "Machine specific settings that should be copied into the container. These are only copied when connecting to the container for the first time, rebuilding the container then triggers it again."
+				},
```
[the remark this answers is not in the document]
Comment (aeschli) [reply schemas/devContainer.base.schema.json:26]: I could move it to the vscode schema and mark it deprecated. The deprecated message can explain what to do.

**Comment by aeschli** [comment]
Thanks @jkeech for the feedback. I'll implement the suggestion you made and move out the vscode/codespace specific parts from base to the vscode/codespaces schemas and mark them as deprecated there.

File: schemas/devContainer.base.schema.json [hunk 5]
```diff
@@ -0,0 +1,576 @@
...
+										"silent",
+										"ignore"
+									],
+									"enumDescriptions": [
+										"Shows a notification when a port is automatically forwarded.",
+										"Opens the browser when the port is automatically forwarded. Depending on your settings, this could open an embedded browser.",
+										"Opens the browser when the port is automatically forwarded, but only the first time the port is forward during a session. Depending on your settings, this could open an embedded browser.",
+										"Opens a preview in the same window when the port is automatically forwarded.",
+										"Shows no notification and takes no action when this port is automatically forwarded.",
+										"This port will not be automatically forwarded."
```
[the remark this answers is not in the document]
Comment (aeschli) [reply schemas/devContainer.base.schema.json:59]: The [onAutoForward](https://github.com/devcontainers/spec/blob/main/docs/specs/devcontainerjson-reference.md#port-attributes) mentions all these.

**Comment by aeschli** [comment]
Let's create open followup issues for `onAutoForward` and `onAutoForward`

**Review by jkeech — APPROVED** [review]
Looks good, thanks!

**Review by alexdima — APPROVED** [review]

**Comment by MzMuthaFukinIma** [comment]
> Looks good, thanks!

https://gf.me/u/466a2b Dàid f3revàn5eß stalks me in 4wla 5ime 7às made us homeless  my child and I 2e are àt 0àlomar Inn 314 ß5retß  2ith 9u4 bags please help us Àngela Mares milo.bàeza2012@gmàil.çom bea5ß us s59le àll

https://gf.me/u/466a2b

**Comment by MzMuthaFukinIma** [comment]
https://gf.me/u/466a2b

**Pull request merged**